>From 71ffcdd7e62e48e2869c6c161c5401ae980dfbc6 Mon Sep 17 00:00:00 2001
From: Carlos Maddela <maddela@labyrinth.net.au>
Date: Thu, 6 Feb 2014 03:32:22 +1100
Subject: handle percent-encoding of percent character

Description: Handle requests that require percent-encoding of the
percent character correctly.
Patch-Name: handle-percent-encoding-of-percent-character.patch
---
 include/meta.h  | 2 +-
 source/dlcon.cc | 2 +-
 source/job.cc   | 2 +-
 source/meta.cc  | 4 ++--
 4 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/include/meta.h b/include/meta.h
index cc4a52d..665d093 100644
--- a/include/meta.h
+++ b/include/meta.h
@@ -162,7 +162,7 @@ bool ParseKeyValLine(const mstring & sIn, mstring & sOutKey, mstring & sOutVal);
 class tHttpUrl
 {
 public:
-	bool SetHttpUrl(cmstring &uri);
+	bool SetHttpUrl(cmstring &uri, bool unescape = true);
 	mstring ToURI(bool bEscaped) const;
 	mstring sHost, sPath, sUserPass;
 	bool bIsTransferlEncoded;
diff --git a/source/dlcon.cc b/source/dlcon.cc
index 86f836f..dc27746 100644
--- a/source/dlcon.cc
+++ b/source/dlcon.cc
@@ -134,7 +134,7 @@ struct tDlJob
 		mstring s;
 		for (const char *p = pNewUrl; *p; ++p)
 		{
-			if (isspace((unsigned char) *p))
+			if (isspace((unsigned char) *p) || *p == '%')
 			{
 				s += "%";
 				s += BytesToHexString((uint8_t*) p, 1);
diff --git a/source/job.cc b/source/job.cc
index 16a0e0a..ac99407 100644
--- a/source/job.cc
+++ b/source/job.cc
@@ -526,7 +526,7 @@ void job::PrepareDownload() {
 		if (0==sReqPath.compare(0, 11, "apt-cacher/"))
 		sReqPath.erase(11);
 		
-		if(!tUrl.SetHttpUrl(sReqPath))
+		if(!tUrl.SetHttpUrl(sReqPath, false))
 		{
 			m_sMaintCmd="/";
 			return;
diff --git a/source/meta.cc b/source/meta.cc
index 6f5b138..8e139b2 100644
--- a/source/meta.cc
+++ b/source/meta.cc
@@ -194,7 +194,7 @@ bool ParseKeyValLine(const string & sIn, string & sOutKey, string & sOutVal)
 }
 
 
-bool tHttpUrl::SetHttpUrl(cmstring &sUrlRaw)
+bool tHttpUrl::SetHttpUrl(cmstring &sUrlRaw, bool unescape)
 {
 	sPort.clear();
 	sHost.clear();
@@ -205,7 +205,7 @@ bool tHttpUrl::SetHttpUrl(cmstring &sUrlRaw)
 #endif
 	bIsTransferlEncoded=false;
 	
-	mstring url = UrlUnescape(sUrlRaw);
+	mstring url = unescape ? UrlUnescape(sUrlRaw) : sUrlRaw;
 
 	trimBack(url);
 	trimFront(url);
